strapi/strapi
# [CTB] fix: preserve pluginOptions in relation properties

### What does it do?

Preserve `pluginOptions` in relation attributes, that are set through the admin interface or manually in the content type schemas.

Preserve in this context means:
- if the options are added/changed in the frontend, they are saved to disk
- if options are already present in an updated relation attribute, they are not deleted
- if options are present in the opposite relation, they are not deleted

### Why is it needed?

When trying to add custom attributes to relation attributes through a plugin, currently all values are deleted before the new schema is saved to disk.

It is necessary to support this use case of adding `pluginOptions` to relation attributes, since the `ctbFormsAPI` suggests that it is possible.

Since the relation attributes are very strict on which values can be set, it should be a reasonable approach to include pluginOptions explicitely instead of preserving all properties.

### How to test it?

Create a simple field extension for relations (in a plugin or admin boostrap):

```js
bootstrap(app) {
    const ctbPlugin = app.getPlugin('content-type-builder')

    if (ctbPlugin) {
      const ctbFormsAPI = ctbPlugin.apis.forms

      ctbFormsAPI.extendFields(['relation'], {
        validator: () => {},
        form: {
          advanced({
            contentTypeSchema,
            forTarget,
            type,
            step,
            data,
            ...rest
          }) {

            return [
              {
                name: 'pluginOptions.myplugin.example',
                type: 'select',
                intlLabel: {
                  id: getTrad('content-type-builder.form.label'),
                  defaultMessage:
                    'Configure automated translation for this field?',
                },
                description: {
                  id: getTrad('content-type-builder.form.description'),
                  defaultMessage:
                    'How should the DeepL plugin handle the translation of this field?',
                },
                validations: {},
                options: [
                  {
                    key: '__null_reset_value__',
                    value: '',
                    metadatas: {
                      intlLabel: {
                        id: 'components.InputSelect.option.placeholder',
                        defaultMessage: 'Choose here',
                      },
                    },
                  },
                  {
                    key: 'first',
                    value: 'first',
                    metadatas: {
                      intlLabel: {
                        id: getTrad(
                          'content-type-builder.form.value.first'
                        ),
                        defaultMessage: 'First',
                      },
                    },
                  },
                  {
                    key: 'second',
                    value: 'second',
                    metadatas: {
                      intlLabel: {
                        id: getTrad('content-type-builder.form.value.second'),
                        defaultMessage: 'Second',
                      },
                    },
                  },
                ],
              },
            ]
          },
        },
      })
    }
  },
```

Then using the `examples/getstarted`, try to set the custom Setting on a relation attribute (e.g. Address > categories) to any value.

Without this merge request, the behaviour is the following:
- Set the custom setting and press finish
- Reopen the relations custom settings and see that it is gone


With this merge request, the behaviour is the following:

One side:
- Set the custom setting and press finish
- Save the content type
- Reopen the relations custom settings and see that it is still there

Both Sides:
- Set the custom setting on the opposite relation and press finish
- Save the content type
- Open the original content types relations custom settings and see that it is still there

### Related issue(s)/PR(s)

Let us know if this is related to any issue/pull request
